### PR TITLE
Fix concurrency issue

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Found some concurrency issues while adding logging to Kestrel

### CODE REVIEW

File: src/Microsoft.Extensions.Logging.Testing/TestSink.cs [hunk 1]
```diff
@@ -31,15 +33,21 @@ public void Write(WriteContext context)
         {
             if (WriteEnabled == null || WriteEnabled(context))
             {
-                Writes.Add(context);
+                lock(_writeLock)
```
Comment (pakrym) [inline src/Microsoft.Extensions.Logging.Testing/TestSink.cs]: ` `

**Comment by JunTaoLuo** [comment]
Apparently this fixes #797

**Comment by JunTaoLuo** [comment]
updated the implementation

File: test/Microsoft.Extensions.Logging.Test/LoggerExtensionsTest.cs [hunk 2]
```diff
@@ -501,47 +501,47 @@ public void LogValuesAndError_LogsCorrectValues()
...
+            sink.Writes.TryDequeue(out var critical);
             Assert.Equal(LogLevel.Critical, critical.LogLevel);
             Assert.Equal(0, critical.EventId);
             Assert.Equal(_exception, critical.Exception);
             Assert.Equal(
                 "Test 1",
                 critical.Formatter(critical.State, critical.Exception));
 
-            var debug = sink.Writes[5];
+            sink.Writes.TryDequeue(out var debug);
```
Comment (pakrym) [inline test/Microsoft.Extensions.Logging.Test/LoggerExtensionsTest.cs]: Assert.True(

File: src/Microsoft.Extensions.Logging.Testing/ITestSink.cs [hunk 3]
```diff
@@ -9,9 +10,9 @@ public interface ITestSink
 
         Func<BeginScopeContext, bool> BeginEnabled { get; set; }
 
-        List<BeginScopeContext> Scopes { get; set; }
+        ConcurrentQueue<BeginScopeContext> Scopes { get; set; }
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Logging.Testing/ITestSink.cs]: ConcurrentQueue

**Comment by JamesNK** [comment]
Will this change eliminate the need for this provider? https://github.com/aspnet/SignalR/blob/dev/test/Microsoft.AspNetCore.SignalR.Tests.Utils/ServerFixture.cs#L148-L206

**Comment by JamesNK** [comment]
// @anurse

**Comment by JunTaoLuo** [comment]
Though mostly the same, LogRecord also contains the date time which isn't included on the WriteContext. I think there's ways to work around it though with formatters and states.

**Comment by pakrym** [comment]
> Can it not be added to WriteContext? I don't remember the "ship/noship" state of this package :)

It can, what do you use time for?

File: test/Microsoft.Extensions.Logging.Test/LoggerExtensionsTest.cs [hunk 4]
```diff
@@ -40,46 +41,48 @@ public void MessageOnly_LogsCorrectValues()
             logger.LogDebug(_state);
             logger.Log(LogLevel.Information, _state);
 
+            var writeList = sink.Writes.ToList();
```
Comment (Tratcher) [inline test/Microsoft.Extensions.Logging.Test/LoggerExtensionsTest.cs]: Do you expect tests elsewhere to be written like this? Or should they be popping items off the front of the list?

Comment (JunTaoLuo) [reply test/Microsoft.Extensions.Logging.Test/LoggerExtensionsTest.cs]: Only if you need to check the order. I see most of the tests use something like Assert.Empty(sink.Writes), Assert.Single(sink.Writes, predicate) or Assert.Contains(sink.Writes, w => predicate) which works with just the enumerable interface. Enumerating as a list is the simplest for test back compat for now but using the TryTake method could also work.

Comment (Tratcher) [reply test/Microsoft.Extensions.Logging.Test/LoggerExtensionsTest.cs]: This test should reflect our recommendations for other repos.

Comment (pakrym) [reply test/Microsoft.Extensions.Logging.Test/LoggerExtensionsTest.cs]: I liked the TryTake version of this, much easier to maintain.

File: src/Microsoft.Extensions.Logging.Testing/ITestSink.cs [hunk 5]
```diff
@@ -9,9 +9,9 @@ public interface ITestSink
 
         Func<BeginScopeContext, bool> BeginEnabled { get; set; }
 
-        List<BeginScopeContext> Scopes { get; set; }
+        IProducerConsumerCollection<BeginScopeContext> Scopes { get; set; }
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Logging.Testing/ITestSink.cs]: 👍

**Pull request merged**